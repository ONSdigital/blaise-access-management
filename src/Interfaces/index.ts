export interface UploadedUser {
    name: string
    created: boolean
}

export interface ImportUser {
    name: string
    password: string
    role: string
    valid: boolean
    warnings: string[]
}

type JSONValue =
    | string
    | { [x: string]: JSONValue };

export interface JSONObject {
    [x: string]: JSONValue;
}

export interface UserForm {
    username?: string
    password?: string
    confirm_password?: string
}

export interface ValidatorType { name: string, validators: ((val: string, name: string, formData: { [key: string]: string }) => string[])[] }

export interface ContextProviderType {
    errors: {
        [key: string]: string[];
    };
    data: {
        [key: string]: string;
    };
    setFieldValue: (name: string, value: string) => void;
    registerInput: ({ name, validators }: ValidatorType) => () => void;
}

export interface BreadcrumbItem {
    link: string
    title: string
}

export interface BreadcrumbProps {
    BreadcrumbList: BreadcrumbItem[]
}