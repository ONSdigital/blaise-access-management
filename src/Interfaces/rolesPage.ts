import { UserRole } from "blaise-api-node-client";

export interface RolesTableProps {
    roles: UserRole[];
    listError: string;
}