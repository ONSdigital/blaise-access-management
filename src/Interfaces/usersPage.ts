import { User } from "blaise-api-node-client";
import { ImportUser, UploadedUser } from ".";

export interface UserRouteParams {
    user: string;
}

export interface UsersProps {
    currentUser: User | undefined;
}

export interface UsersTableProps {
    users: User[];
    currentUser: User | undefined;
    listError: string;
}

export interface SelectFileProps {
    setUsersToUpload: (users: ImportUser[]) => void;
    movePageForward: () => void;
}

export interface UsersToUploadSummaryProps {
    usersToImport: ImportUser[]
    uploadUsers: () => void
}

export interface UsersUploadedSummaryProps {
    usersUploaded: UploadedUser[]
    numberOfValidUsers: number
}

export interface ReturnPanel {
    visible: boolean
    message: string
    status: string
}