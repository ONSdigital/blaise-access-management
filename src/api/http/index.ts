import axios from "axios";

export * from "./users";
export * from "./roles";

export function handleAxiosError(error: unknown): { status: number; message: string, error: unknown } {
    if (axios.isAxiosError(error)) {
        return {
            status: error.response?.status || 500,
            message: error.response?.data.message || "Axios error occurred with no specific message",
            error: error
        };
    } else {
        return {
            status: 500,
            message: "An unknown error occurred, please raise a ServiceNow call if this continues: " + error,
            error: error
        };
    }
}