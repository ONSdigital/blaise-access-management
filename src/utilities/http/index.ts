export * from "./users";
export * from "./roles";
export * from "./auth";
