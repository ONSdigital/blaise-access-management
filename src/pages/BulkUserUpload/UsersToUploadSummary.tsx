import React, {ReactElement, useEffect, useState} from "react";
import {ONSPanel} from "blaise-design-system-react-components";
import {ImportUser} from "../../../Interfaces";
import {ErrorBoundary} from "../../Components/ErrorHandling/ErrorBoundary";
import Confirmation from "./Confirmation";
import converter from "number-to-words";

interface Props {
    statusDescriptionList: ImportUser[]
    uploadUsers: () => void
}

function UsersToUploadSummary({statusDescriptionList, uploadUsers}: Props): ReactElement {
    const [userList, setUserList] = useState<ImportUser[]>([]);
    const [listError, setListError] = useState<string>("Loading ...");
    const [noValidUsers, setNoValidUsers] = useState<number>(0);

    useEffect(() => {
        console.log("UsersToUploadSummary started");

        console.log(userList);

        setupUserList().then(() => console.log("SetupUserList complete"));
    }, [statusDescriptionList]);

    async function setupUserList() {
        setListError("Loading ...");

        let noValid = 0;
        statusDescriptionList.map((user: ImportUser) => {
            if (user.valid) {
                noValid = noValid + 1;
            }
        });

        setNoValidUsers(noValid);
        statusDescriptionList.sort((a, b) => (a.valid ? 1 : 0) - (b.valid ? 1 : 0));

        if (statusDescriptionList.length === 0) {
            setListError("No users found to upload");
        }

        await setUserList(statusDescriptionList);
    }


    return (
        <>
            <h1 className="u-mt-l">Bulk upload <em>{converter.toWords(noValidUsers)}</em> user{(noValidUsers > 1 && "s")}?</h1>
            <ONSPanel>
                <p>{noValidUsers} of {userList.length} users are valid and will be uploaded. <em>Invalid users will not be uploaded.
                </em> You can review any issues in the table below.</p>
            </ONSPanel>

            <Confirmation validUsers={noValidUsers} uploadUsers={uploadUsers}/>
            <h2 className="u-mt-xl">Users to upload</h2>
            <ErrorBoundary errorMessageText={"Failed to load audit logs."}>
                {
                    userList && userList.length > 0
                        ?
                        <table id="batch-table" className="table">
                            <thead className="table__head u-mt-m">
                            <tr className="table__row">
                                <th scope="col" className="table__header ">
                                    <span>Username</span>
                                </th>
                                <th scope="col" className="table__header ">
                                    <span>Role</span>
                                </th>
                                <th scope="col" className="table__header ">
                                    <span>User validity</span>
                                </th>
                            </tr>
                            </thead>
                            <tbody className="table__body">
                            {
                                userList.map(({name, role, valid, warnings}: ImportUser, index: number) => {

                                    return (
                                        <tr className="table__row" key={name + index}
                                            data-testid={"batch-table-row"}>

                                            <td className="table__cell ">
                                                {name}
                                            </td>
                                            <td className="table__cell ">
                                                {role}
                                            </td>
                                            <td className="table__cell ">
                                                <span className={`status status--${(valid ? "success" : "error")}`}>
                                                    {
                                                        valid
                                                            ? "Valid User"
                                                            : warnings.map((message) => {
                                                                return (`${message}. `);
                                                            })
                                                    }
                                                </span>
                                            </td>

                                        </tr>
                                    );
                                })
                            }
                            </tbody>
                        </table>
                        :
                        <ONSPanel>{listError}</ONSPanel>
                }
            </ErrorBoundary>
        </>
    );
}

export default UsersToUploadSummary;
