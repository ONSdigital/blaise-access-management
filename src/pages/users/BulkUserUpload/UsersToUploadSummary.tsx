import React, {ReactElement, useEffect, useState} from "react";
import {ErrorBoundary, ONSPanel} from "blaise-design-system-react-components";
import {ImportUser} from "../../../../Interfaces";
import Confirmation from "./Confirmation";
import converter from "number-to-words";
import pino from "pino";

const logger = pino();

interface Props {
    statusDescriptionList: ImportUser[]
    uploadUsers: () => void
}

function UsersToUploadSummary({statusDescriptionList, uploadUsers}: Props): ReactElement {
    const [userList, setUserList] = useState<ImportUser[]>([]);
    const [listError, setListError] = useState<string>("Loading ...");
    const [noValidUsers, setNoValidUsers] = useState<number>(0);

    useEffect(() => {
        setupUserList().then(() => logger.info("SetupUserList is complete"));
    }, [statusDescriptionList]);

    async function setupUserList() {
        setListError("Loading ...");

        let noValid = 0;
        statusDescriptionList.map((user: ImportUser) => {
            if (user.valid) {
                noValid = noValid + 1;
            }
        });

        setNoValidUsers(noValid);
        statusDescriptionList.sort((a, b) => (a.valid ? 1 : 0) - (b.valid ? 1 : 0));

        if (statusDescriptionList.length === 0) {
            setListError("No users found to upload");
        }

        await setUserList(statusDescriptionList);
    }

    return (
        <>
            <h1 className="ons-u-mb-l">Bulk upload <em>{converter.toWords(noValidUsers)}</em> user{(noValidUsers > 1 && "s")}?</h1>
            <ONSPanel>
                <p>{noValidUsers} of {userList.length} users are valid and will be uploaded. <em>Invalid users will not be uploaded.
                </em> You can review any issues in the table below.</p>
            </ONSPanel>

            <Confirmation validUsers={noValidUsers} uploadUsers={uploadUsers}/>
            <h2 className="ons-u-mt-xl">Users to upload</h2>
            <ErrorBoundary errorMessageText={"Failed to load audit logs."}>
                {
                    userList && userList.length > 0
                        ?
                        <table id="batch-table" className="ons-table">
                            <thead className="ons-table__head ons-u-mt-m">
                            <tr className="ons-table__row">
                                <th scope="col" className="ons-table__header ">
                                    <span>Username</span>
                                </th>
                                <th scope="col" className="ons-table__header ">
                                    <span>Role</span>
                                </th>
                                <th scope="col" className="ons-table__header ">
                                    <span>User validity</span>
                                </th>
                            </tr>
                            </thead>
                            <tbody className="ons-table__body">
                            {
                                userList.map(({name, role, valid, warnings}: ImportUser, index: number) => {

                                    return (
                                        <tr className="ons-table__row" key={name + index}
                                            data-testid={"batch-table-row"}>

                                            <td className="ons-table__cell ">
                                                {name}
                                            </td>
                                            <td className="ons-table__cell ">
                                                {role}
                                            </td>
                                            <td className="ons-table__cell ">
                                                <span className={`ons-status ons-status--${(valid ? "success" : "error")}`}>
                                                    {
                                                        valid
                                                            ? "Valid User"
                                                            : warnings.map((message) => {
                                                                return (`${message}. `);
                                                            })
                                                    }
                                                </span>
                                            </td>

                                        </tr>
                                    );
                                })
                            }
                            </tbody>
                        </table>
                        :
                        <ONSPanel>{listError}</ONSPanel>
                }
            </ErrorBoundary>
        </>
    );
}

export default UsersToUploadSummary;
