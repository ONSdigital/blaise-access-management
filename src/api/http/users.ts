import axios from "axios";
import { requestPromiseJson, requestPromiseJsonList } from "./requestPromise";
import { User, NewUser } from "blaise-api-node-client";
import { AuthManager } from "blaise-login-react/blaise-login-react-client";
import { GetUserResponse, GetUsersListResponse, PatchUserRoleResponse } from "../../Interfaces/usersPage";
import { handleAxiosError } from ".";

async function getAllUsers(): Promise<GetUsersListResponse> {
    try {
        const url = "/api/users";
        const [success, data] = await requestPromiseJsonList<User>("GET", url);
        return [success, data] as GetUsersListResponse;
    } catch (error) {
        return [false, []] as GetUsersListResponse;
    }
}

// NOTE: Catch block is different due to the use of AsyncContent and AsyncRequest
async function getUser(user: string): Promise<GetUserResponse> {
    try {
        const authManager = new AuthManager();
        const url = `/api/users/${user}`;
        const headers = authManager.authHeader();

        const res = await axios.get(url, { headers });

        return {
            status: res.status,
            message: res.data.message,
            data: res.data.data
        };

    } catch (error) {
        const errorObj = handleAxiosError(error);
        console.error(errorObj.message, { ...errorObj, data: {}, error: error });
        throw Error(errorObj.message);
    }
}

function addNewUser(newUser: NewUser): Promise<boolean> {
    const url = "/api/users";

    return new Promise((resolve: (object: boolean) => void) => {

        if (!newUser.password) {
            resolve(false);
            return;
        }

        const formData = new FormData();
        formData.append("name", newUser.name);
        formData.append("password", newUser.password);
        formData.append("role", newUser.role);

        requestPromiseJson("POST", url, formData).then(([status]) => {
            if (status === 200 || status === 201) {
                resolve(true);
            } else {
                resolve(false);
            }
        }).catch(() => {
            resolve(false);
        });
    });
}

function deleteUser(username: string): Promise<boolean> {
    const url = "/api/users";

    const headers = {
        "user": username
    };

    return new Promise((resolve: (object: boolean) => void) => {

        requestPromiseJson("DELETE", url, null, headers).then(([status]) => {
            if (status === 204) {
                resolve(true);
            } else {
                resolve(false);
            }
        }).catch(() => {
            resolve(false);
        });
    });
}

// BLAIS5-3816
async function patchUserRolesAndPermissions(user: string, role: string, currentServerParks: string[], currentDefaultServerPark: string): Promise<PatchUserRoleResponse> {
    try {
        const authManager = new AuthManager();
        const url = `/api/users/${user}/rolesAndPermissions`;
        const headers = authManager.authHeader();
        const data = { role, currentServerParks, currentDefaultServerPark };

        const res = await axios.patch(url, data, { headers });

        return {
            status: res.status,
            message: res.data.message
        };
    } catch (error) {
        const errorObj = handleAxiosError(error);
        console.error(errorObj.message, errorObj);
        return errorObj;
    }
}

export { getAllUsers, getUser, addNewUser, deleteUser, patchUserRolesAndPermissions };
