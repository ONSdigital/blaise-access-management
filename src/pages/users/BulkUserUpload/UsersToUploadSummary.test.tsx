import React from "react";
import {render, waitFor, cleanup, screen, RenderResult} from "@testing-library/react";
import "@testing-library/jest-dom";
import {act} from "react-dom/test-utils";
import {createMemoryHistory} from "history";
import {Router} from "react-router";
import { ImportUser } from "../../../../Interfaces";
import UsersToUploadSummary from "./UsersToUploadSummary";
import { getAllRoles } from "../../../utilities/http";
import { UserRole } from "blaise-api-node-client";

// set global vars 
let view:RenderResult;

// set mocks
type getRolesListResponse = [boolean, UserRole[]];
jest.mock('../../../utilities/http');
const getAllRolesMock = getAllRoles as jest.Mock<Promise<getRolesListResponse>>;

describe("Upload summary tests", () => {

    var roles: getRolesListResponse = [
        true, 
        [
        { 
            name: "BDSS", 
            description: "",
            permissions: []
        },
        { 
            name: "DST", 
            description: "",
            permissions: []
        }            
    ]]

    beforeEach(() => {
        getAllRolesMock.mockImplementation(() => Promise.resolve(roles));
      });


    it("view users page matches Snapshot", async () => {
        //arrange
        var userList: ImportUser[] = [
            {
                name:'Jamie',
                password:'pass',
                role:'BDSS',
                valid:false,
                warnings:[]
            },
            {
                name:'Rob',
                password:'pass2',
                role:'DST',
                valid:false,
                warnings:[]
            },    
            {
                name:'Jamie',
                password:'pass',
                role:'BDSS',
                valid:false,
                warnings:[]
            },       
            ]

        const history = createMemoryHistory();
        const view = render(
            <Router history={history}>
                <UsersToUploadSummary usersToImport={userList} uploadUsers={() => {}}/>
            </Router>
        );

        // act
        await act(async () => {
            await new Promise(process.nextTick);
        });

        // assert
        await waitFor(() => {
            expect(view).toMatchSnapshot();
        });
    });

    it("Upload summary pages for three valid users matches Snapshot", async () => {
        //arrange
        var userList: ImportUser[] = [
            {
                name:'Jamie',
                password:'pass',
                role:'BDSS',
                valid:false,
                warnings:[]
            },
            {
                name:'Rob',
                password:'pass2',
                role:'DST',
                valid:false,
                warnings:[]
            },    
            {
                name:'Rich',
                password:'pass',
                role:'BDSS',
                valid:false,
                warnings:[]
            },       
            ]

        // act
        await act(async () => {
            const history = createMemoryHistory();
            const view = render(
                <Router history={history}>
                    <UsersToUploadSummary usersToImport={userList} uploadUsers={() => {}}/>
                </Router>
            );
        });

        // assert
        await waitFor(() => {
            expect(view).toMatchSnapshot();
        });

    });    

    it("Upload summary pages for three valid users displays correct summary", async () => {
        //arrange
        var userList: ImportUser[] = [
            {
                name:'Jamie',
                password:'pass',
                role:'BDSS',
                valid:false,
                warnings:[]
            },
            {
                name:'Rob',
                password:'pass2',
                role:'DST',
                valid:false,
                warnings:[]
            },    
            {
                name:'Rich',
                password:'pass',
                role:'BDSS',
                valid:false,
                warnings:[]
            },       
            ]

        // act
        // act
        await act(async () => {
            const history = createMemoryHistory();
            const view = render(
                <Router history={history}>
                    <UsersToUploadSummary usersToImport={userList} uploadUsers={() => {}}/>
                </Router>
            );
        });

        // assert
        expect(screen.getByText(/3 of 3 users are valid and will be uploaded/i)).toBeDefined();        
    });    

    afterAll(() => {
        jest.clearAllMocks();
        cleanup();
    });
});

