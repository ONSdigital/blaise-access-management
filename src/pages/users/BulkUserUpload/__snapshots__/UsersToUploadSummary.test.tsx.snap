// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload summary tests Upload summary pages for three valid users matches Snapshot 1`] = `undefined`;

exports[`Upload summary tests view users page matches Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h1
        class="ons-u-mb-l"
      >
        Bulk upload 
        <em>
          three
        </em>
         user
        s
        ?
      </h1>
      <div
        class="ons-panel ons-panel--info ons-panel--no-title  ons-u-mt-m"
        data-testid="summary-panel"
      >
        <div
          class="ons-panel__body "
        >
          <p>
            3
             of 
            3
             users are valid and will be uploaded. 
            <em>
              Invalid users will not be uploaded.
            </em>
             You can review any issues in the table below.
          </p>
        </div>
      </div>
      <form
        class="ons-u-mt-s"
      >
        <fieldset
          class="ons-fieldset"
        >
          <legend
            class="ons-fieldset__legend"
          />
          <div
            class="ons-radios__items"
          >
            <p
              class="ons-radios__item"
            >
              <span
                class="ons-radio"
              >
                <input
                  aria-label="Yes"
                  class="ons-radio__input ons-js-radio "
                  id="confirm-upload"
                  name="confirm-upload"
                  type="radio"
                  value="True"
                />
                <label
                  class="ons-radio__label "
                  for="confirm-upload"
                >
                  Yes, upload 
                  three
                   valid user
                  s
                </label>
              </span>
            </p>
            <br />
            <p
              class="ons-radios__item"
            >
              <span
                class="ons-radio"
              >
                <input
                  aria-label="No"
                  class="ons-radio__input ons-js-radio "
                  id="cancel-upload"
                  name="confirm-upload"
                  type="radio"
                  value="False"
                />
                <label
                  class="ons-radio__label "
                  for="cancel-upload"
                >
                  No, do not upload any users
                </label>
              </span>
            </p>
          </div>
        </fieldset>
        <br />
        <button
          class="ons-btn "
          data-testid="button"
          id="confirm-continue"
          type="button"
        >
          <span
            class="ons-btn__inner"
          >
            Continue
          </span>
        </button>
        <button
          class="ons-btn  ons-btn--secondary"
          data-testid="button"
          id="cancel-overwrite"
          type="button"
        >
          <span
            class="ons-btn__inner"
          >
            Cancel
          </span>
        </button>
      </form>
      <h2
        class="ons-u-mt-xl"
      >
        Users to upload
      </h2>
      <table
        class="ons-table"
        id="batch-table"
      >
        <thead
          class="ons-table__head ons-u-mt-m"
        >
          <tr
            class="ons-table__row"
          >
            <th
              class="ons-table__header "
              scope="col"
            >
              <span>
                Username
              </span>
            </th>
            <th
              class="ons-table__header "
              scope="col"
            >
              <span>
                Role
              </span>
            </th>
            <th
              class="ons-table__header "
              scope="col"
            >
              <span>
                User validity
              </span>
            </th>
          </tr>
        </thead>
        <tbody
          class="ons-table__body"
        >
          <tr
            class="ons-table__row"
            data-testid="batch-table-row"
          >
            <td
              class="ons-table__cell "
            >
              Jamie
            </td>
            <td
              class="ons-table__cell "
            >
              BDSS
            </td>
            <td
              class="ons-table__cell "
            >
              <span
                class="ons-status ons-status--success"
              >
                Valid User
              </span>
            </td>
          </tr>
          <tr
            class="ons-table__row"
            data-testid="batch-table-row"
          >
            <td
              class="ons-table__cell "
            >
              Rob
            </td>
            <td
              class="ons-table__cell "
            >
              DST
            </td>
            <td
              class="ons-table__cell "
            >
              <span
                class="ons-status ons-status--success"
              >
                Valid User
              </span>
            </td>
          </tr>
          <tr
            class="ons-table__row"
            data-testid="batch-table-row"
          >
            <td
              class="ons-table__cell "
            >
              Jamie
            </td>
            <td
              class="ons-table__cell "
            >
              BDSS
            </td>
            <td
              class="ons-table__cell "
            >
              <span
                class="ons-status ons-status--success"
              >
                Valid User
              </span>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <h1
      class="ons-u-mb-l"
    >
      Bulk upload 
      <em>
        three
      </em>
       user
      s
      ?
    </h1>
    <div
      class="ons-panel ons-panel--info ons-panel--no-title  ons-u-mt-m"
      data-testid="summary-panel"
    >
      <div
        class="ons-panel__body "
      >
        <p>
          3
           of 
          3
           users are valid and will be uploaded. 
          <em>
            Invalid users will not be uploaded.
          </em>
           You can review any issues in the table below.
        </p>
      </div>
    </div>
    <form
      class="ons-u-mt-s"
    >
      <fieldset
        class="ons-fieldset"
      >
        <legend
          class="ons-fieldset__legend"
        />
        <div
          class="ons-radios__items"
        >
          <p
            class="ons-radios__item"
          >
            <span
              class="ons-radio"
            >
              <input
                aria-label="Yes"
                class="ons-radio__input ons-js-radio "
                id="confirm-upload"
                name="confirm-upload"
                type="radio"
                value="True"
              />
              <label
                class="ons-radio__label "
                for="confirm-upload"
              >
                Yes, upload 
                three
                 valid user
                s
              </label>
            </span>
          </p>
          <br />
          <p
            class="ons-radios__item"
          >
            <span
              class="ons-radio"
            >
              <input
                aria-label="No"
                class="ons-radio__input ons-js-radio "
                id="cancel-upload"
                name="confirm-upload"
                type="radio"
                value="False"
              />
              <label
                class="ons-radio__label "
                for="cancel-upload"
              >
                No, do not upload any users
              </label>
            </span>
          </p>
        </div>
      </fieldset>
      <br />
      <button
        class="ons-btn "
        data-testid="button"
        id="confirm-continue"
        type="button"
      >
        <span
          class="ons-btn__inner"
        >
          Continue
        </span>
      </button>
      <button
        class="ons-btn  ons-btn--secondary"
        data-testid="button"
        id="cancel-overwrite"
        type="button"
      >
        <span
          class="ons-btn__inner"
        >
          Cancel
        </span>
      </button>
    </form>
    <h2
      class="ons-u-mt-xl"
    >
      Users to upload
    </h2>
    <table
      class="ons-table"
      id="batch-table"
    >
      <thead
        class="ons-table__head ons-u-mt-m"
      >
        <tr
          class="ons-table__row"
        >
          <th
            class="ons-table__header "
            scope="col"
          >
            <span>
              Username
            </span>
          </th>
          <th
            class="ons-table__header "
            scope="col"
          >
            <span>
              Role
            </span>
          </th>
          <th
            class="ons-table__header "
            scope="col"
          >
            <span>
              User validity
            </span>
          </th>
        </tr>
      </thead>
      <tbody
        class="ons-table__body"
      >
        <tr
          class="ons-table__row"
          data-testid="batch-table-row"
        >
          <td
            class="ons-table__cell "
          >
            Jamie
          </td>
          <td
            class="ons-table__cell "
          >
            BDSS
          </td>
          <td
            class="ons-table__cell "
          >
            <span
              class="ons-status ons-status--success"
            >
              Valid User
            </span>
          </td>
        </tr>
        <tr
          class="ons-table__row"
          data-testid="batch-table-row"
        >
          <td
            class="ons-table__cell "
          >
            Rob
          </td>
          <td
            class="ons-table__cell "
          >
            DST
          </td>
          <td
            class="ons-table__cell "
          >
            <span
              class="ons-status ons-status--success"
            >
              Valid User
            </span>
          </td>
        </tr>
        <tr
          class="ons-table__row"
          data-testid="batch-table-row"
        >
          <td
            class="ons-table__cell "
          >
            Jamie
          </td>
          <td
            class="ons-table__cell "
          >
            BDSS
          </td>
          <td
            class="ons-table__cell "
          >
            <span
              class="ons-status ons-status--success"
            >
              Valid User
            </span>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
